### PR TITLE
refactor pillars and added yaml decode

### PR DESCRIPTION
**Pull request by afori**
The limitation of supporting only strings and their placing under parameters make it impossible to use any salt formula.
This pull request:
- refactors pillars moving all foreman specific under "foreman"
- decodes a yaml if a parameter starts with "---\r\n"
- moves the decoded paramaters under root

Use case:
formula: https://github.com/saltstack-formulas/iptables-formula
add parameter forewall="---\r\nfirewall:\r\n  enabled: True\r\n  install: True\r\n  strict: True...."
Note that the full screen editor for paramaters makes very easy to edit paramateres in yaml.

### CODE REVIEW

**Comment by stbenjam** [comment]
Hi there!

Thanks for the PR.

> refactors pillars moving all foreman specific under "foreman"

Why is everything under a 'foreman' key? Salt master_tops and the external pillars expect a very particular format, which is a hash containing `classes` (list of states) and `parameters` (hash of pillars) top level keys.

If you want pillars namespaced to a 'foreman' key, this is already possible, Administer / Settings, and flip the setting to true, but you've got it in the wrong spot.

> decodes a yaml if a parameter starts with "---\r\n"

This is a foreman issue anyway and not something with the plugin, it is tracked by: http://projects.theforeman.org/issues/4127

I realize that reduces the usability of Foreman parameters as pillars, but once the above feature is available it'll just work with the plugin.

In the interim, I might be ok with parameters beginning with '---' interpreted as YAML but I don't think your code works, and it needs to be tested.

File: app/controllers/foreman_salt/minions_controller.rb [hunk 1]
```diff
@@ -6,14 +6,21 @@ class MinionsController < ::ForemanSalt::ApplicationController
...
-      enc['classes'] = @minion.salt_modules_for_enc
+      enc['foreman']['classes'] = @minion.salt_modules_for_enc
 
       pillars = @minion.info['parameters']
-      enc['parameters'] = Setting[:salt_namespace_pillars] ? { 'foreman' => pillars } : pillars
+      enc['foreman']['parameters'] = Setting[:salt_namespace_pillars] ? { 'foreman' => pillars } : pillars
 
-      enc['environment'] = env
+      enc['foreman']['parameters'].each do |key, value|
+        if (value.is_a? String) and value.start_with?("---\r\n")
```
Comment (stbenjam) [inline app/controllers/foreman_salt/minions_controller.rb]: Why `\r\n`? That's a windows thing.

File: app/controllers/foreman_salt/minions_controller.rb [hunk 2]
```diff
@@ -6,14 +6,21 @@ class MinionsController < ::ForemanSalt::ApplicationController
...
+          enc[key] = YAML.load(value)
```
Comment (stbenjam) [inline app/controllers/foreman_salt/minions_controller.rb]: Why under enc[key]? Pillars need to be namespaces under 'parameters' key.

File: app/controllers/foreman_salt/minions_controller.rb [hunk 3]
```diff
@@ -6,14 +6,21 @@ class MinionsController < ::ForemanSalt::ApplicationController
     add_smart_proxy_filters :node, :features => 'Salt'
 
     def node
-      enc = {}
+      enc = {'foreman' => {}}
```
Comment (stbenjam) [inline app/controllers/foreman_salt/minions_controller.rb]: Why? This breaks the foreman-node script, and isn't the data format expected by the salt master.

Comment (afori) [inline app/controllers/foreman_salt/minions_controller.rb]: Sorry, I've misinterpreted how the generated pillars were passed down to the salt master.
Most of salt users like me would expect just to see the pillars as seen by the minion. Now I understand that what is shown pressing "Salt ENC' is the raw data passed to the salt master.
Maybe better to just have a "pillar" button that shows the actual pillar data that minion gets.

Comment (afori) [inline app/controllers/foreman_salt/minions_controller.rb]: BTW We can close a pull request and I can create another one that just render the yaml if string starts with '---'

Comment (stbenjam) [inline app/controllers/foreman_salt/minions_controller.rb]: Ah ok, I see.  Yea, it's using ext_pillar.  We could make that easier to see that's pillars, the underlying format actually comes from Puppet (the idea from the Saltstack devs was to be able to reuse some existing puppet ENC's).

I think that's easy enough to do.  

The YAML change is fine if you want to make the PR just do that, I bet a lot of people would be interested.

Thanks!

File: app/controllers/foreman_salt/minions_controller.rb (hunk 1 above)
Comment (afori) [inline app/controllers/foreman_salt/minions_controller.rb]: Just noticed that if a CR is added after the '---' there is \r\n.
Of course we can just check '---'

**Comment by mmoll** [comment]
@afori what's the status here?

**Pull request closed without merging**